File: registration.py
import threading
import time
import random
import json

class RegistrationSystem:
    def __init__(self, courses_file):
        with open(courses_file, 'r') as f:
            courses_db = json.load(f)
            
        # Keep initial DB state to reset between tests
        self.initial_db = {k: v.copy() for k, v in courses_db.items()}
        self.db = {k: v.copy() for k, v in courses_db.items()}
        
        # Create a mutex lock for each course to protect its seat counter independently
        self.course_locks = {k: threading.Lock() for k in courses_db}
        self.successful_registrations = []

    def reset(self):
        self.db = {k: v.copy() for k, v in self.initial_db.items()}
        self.successful_registrations = []

    def register_unsafe(self, student_id, course_id):
        """
        Registers a student for a course WITHOUT synchronization.
        Demonstrates a race condition.
        """
        seats = self.db[course_id]["seats"]
        
        # Simulate CPU processing or network latency
        time.sleep(random.uniform(0.001, 0.005))
        
        if seats > 0:
            # Simulating another pause where a context switch might occur
            time.sleep(random.uniform(0.001, 0.005))
            self.db[course_id]["seats"] = seats - 1
            self.successful_registrations.append(student_id)
            return True
        return False

    def register_safe(self, student_id, course_id):
        """
        Registers a student using a mutex lock to ensure thread safety.
        """
        with self.course_locks[course_id]:
            seats = self.db[course_id]["seats"]
            
            time.sleep(random.uniform(0.001, 0.005))
            
            if seats > 0:
                time.sleep(random.uniform(0.001, 0.005))
                self.db[course_id]["seats"] = seats - 1
                self.successful_registrations.append(student_id)
                return True
            return False

def simulate_concurrent_registration(system, course_id, num_students, safe=False):
    print(f"\n--- Starting Simulation (Safe Mode: {safe}) ---")
    print(f"Target Course: {course_id}")
    print(f"Initial Seats: {system.db[course_id]['seats']}")
    print(f"Number of Students Trying to Register: {num_students}")
    
    threads = []
    
    for i in range(num_students):
        student_id = f"Student_{i+1}"
        target_function = system.register_safe if safe else system.register_unsafe
        
        t = threading.Thread(target=target_function, args=(student_id, course_id))
        threads.append(t)
        t.start()
        
    for t in threads:
        t.join()
        
    final_seats = system.db[course_id]['seats']
    success_count = len(system.successful_registrations)
    
    print(f"\nSimulation Complete!")
    print(f"Final Seats Remaining: {final_seats}")
    print(f"Total Successful Registrations: {success_count}")
    
    if final_seats < 0:
        print("!!! ERROR: RACE CONDITION DETECTED !!!")
        print("More students registered than available seats! Negative seats encountered.")
    elif success_count > system.initial_db[course_id]['seats']:
        print("!!! ERROR: RACE CONDITION DETECTED !!!")
        print("Successfully registered students exceed initial capacity!")
    else:
        print("System behaved correctly.")

File: registration.js
const fs = require('fs');

// A simple Async Mutex to simulate thread locking in Node's event loop
class AsyncMutex {
    constructor() {
        this.locked = false;
        this.queue = [];
    }

    async acquire() {
        return new Promise(resolve => {
            if (!this.locked) {
                this.locked = true;
                resolve();
            } else {
                this.queue.push(resolve);
            }
        });
    }

    release() {
        if (this.queue.length > 0) {
            const resolve = this.queue.shift();
            resolve();
        } else {
            this.locked = false;
        }
    }
}

class RegistrationSystem {
    constructor(coursesFile) {
        const data = fs.readFileSync(coursesFile, 'utf-8');
        const coursesDb = JSON.parse(data);
        
        this.initialDb = JSON.parse(JSON.stringify(coursesDb));
        this.db = JSON.parse(JSON.stringify(coursesDb));
        
        this.courseLocks = {};
        for (const course in coursesDb) {
            this.courseLocks[course] = new AsyncMutex();
        }
        
        this.successfulRegistrations = [];
    }

    reset() {
        this.db = JSON.parse(JSON.stringify(this.initialDb));
        this.successfulRegistrations = [];
    }

    // A helper to simulate DB latency / CPU time that causes async race conditions
    delay(ms) {
        return new Promise(resolve => setTimeout(resolve, ms));
    }

    async registerUnsafe(studentId, courseId) {
        const seats = this.db[courseId].seats;
        
        // Simulating async gap where other event loop ticks can interleave
        await this.delay(Math.random() * 5 + 1);
        
        if (seats > 0) {
            await this.delay(Math.random() * 5 + 1);
            this.db[courseId].seats = seats - 1;
            this.successfulRegistrations.push(studentId);
            return true;
        }
        return false;
    }

    async registerSafe(studentId, courseId) {
        const lock = this.courseLocks[courseId];
        await lock.acquire(); // Atomic lock check
        
        try {
            const seats = this.db[courseId].seats;
            
            await this.delay(Math.random() * 5 + 1);
            
            if (seats > 0) {
                await this.delay(Math.random() * 5 + 1);
                this.db[courseId].seats = seats - 1;
                this.successfulRegistrations.push(studentId);
                return true;
            }
            return false;
        } finally {
            lock.release();
        }
    }
}

async function simulateConcurrentRegistration(system, courseId, numStudents, safe = false) {
    const promises = [];
    
    for (let i = 0; i < numStudents; i++) {
        const studentId = `Student_${i+1}`;
        if (safe) {
            promises.push(system.registerSafe(studentId, courseId));
        } else {
            promises.push(system.registerUnsafe(studentId, courseId));
        }
    }
    
    await Promise.all(promises);
}

module.exports = { RegistrationSystem, simulateConcurrentRegistration };
